### PR TITLE
An example for a quick start guide

### PR DESCRIPTION
**Pull request by yoni206**
Co-authored-by: Aina Niemetz <aina.niemetz@gmail.com>

### CODE REVIEW

File: examples/api/cpp/quickstart.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  Solver solver;
+
+  // We will ask the solver to produce
+  // models and unsat cores, hence these options
+  // should be turned on.
+  solver.setOption("produce-models", "true");
+  solver.setOption("produce-unsat-cores", "true");
+
+  // The simplest way to set a logic for the solver is to choose "ALL".
+  // This enables all logics in the solver.
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:36]:
```suggestion
  // This enables all logics in the solver.
  // Alternatively, "QF_ALL" enables all logics without quantifiers.
```

File: examples/api/cpp/quickstart.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  // To optimize the solver's behavior for a more specific logic,
+  // use the logic name, e.g. "QF_BV" or "QF_AUFBV".
+
+  // Set the logic
+  solver.setLogic("ALL");
+
+  // In this example, we will define constraints
+  // over reals and integers.
+  // Hence we first obtain the corresponding sorts.
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:45]:
```suggestion
  // Hence, we first obtain the corresponding sorts.
```

File: examples/api/cpp/quickstart.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  Sort realSort = solver.getRealSort();
+  Sort intSort = solver.getIntegerSort();
+
+  // x,y will be real variables, while
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:49]:
```suggestion
  // x and y will be real variables, while
```

File: examples/api/cpp/quickstart.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  // a,b will be integer variables.
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:50]:
```suggestion
  // a and b will be integer variables.
```

File: examples/api/cpp/quickstart.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  // Formally, their cpp type is Term, and
+  // they are called "constants" in SMT jargon:
+  Term x = solver.mkConst(realSort, "x");
+  Term y = solver.mkConst(realSort, "y");
+  Term a = solver.mkConst(intSort, "a");
+  Term b = solver.mkConst(intSort, "b");
+
+  // Our constraints regarding x,y will be:
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:58]:
```suggestion
  // Our constraints regarding x and y will be:
```

File: examples/api/cpp/quickstart.cpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  // '''
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:59]:
```suggestion
  //
```

File: examples/api/cpp/quickstart.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  //   (1)  0 < x
+  //   (2)  0 < y
+  //   (3)  x + y < 1
+  //   (4)  x <= y
+  // '''
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:64]:
```suggestion
  //
```

File: examples/api/cpp/quickstart.cpp [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+  // Formally, constraints are also terms. Their sort is Boolean.
+  // We will construct these constraints gradually,
+  // by defining each of their components.
+  // We start with the constant numerals 0 and 1:
+  Term zero = solver.mkReal(0);
+  Term one = solver.mkReal(1);
+
+  // Next, we construct the term x+y
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:73]:
```suggestion
  // Next, we construct the term x + y
```

File: examples/api/cpp/quickstart.cpp [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  Term xPlusY = solver.mkTerm(PLUS, x, y);
+
+  // Next, we continue with the compound terms, that
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:76]: This looks like a part of the sentence is missing?

File: examples/api/cpp/quickstart.cpp [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+  // Now we can define the constraints.
+  // They employ +, <=, and <.
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:79]:
```suggestion
  // They use the operators +, <=, and <.
```

File: examples/api/cpp/quickstart.cpp [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+  // In the API, these are denoted by
+  // PLUS, LEQ, and LT.
+  // A list of operators is available in:
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:82]:
```suggestion
  // A list of available operators is available in:
```

File: examples/api/cpp/quickstart.cpp [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  Term constraint4 = solver.mkTerm(LEQ, x, y);
+
+  // Now we assert the constraints to the solver.
+  solver.assertFormula(constraint1);
+  solver.assertFormula(constraint2);
+  solver.assertFormula(constraint3);
+  solver.assertFormula(constraint4);
+
+  // Check if the formula is satisfiable, that is,
+  // are there real x,y,z that satisfy all the
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:96]:
```suggestion
  // are there real values for x,y,z that satisfy all the
```

File: examples/api/cpp/quickstart.cpp [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+  // constraints?
+  Result r1 = solver.checkSat();
+
+  // The result is either SAT, UNSAT, or UNKNOWN.
+  // In this case, it is SAT.
+  std::cout << "result:" << r1 << std::endl;
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:102]:
```suggestion
  std::cout << "expected: sat" << std::endl;
  std::cout << "result:" << r1 << std::endl;
```

File: examples/api/cpp/quickstart.cpp [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+  // We can get actual values for x,y that
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:104]:
```suggestion
  // We can get the values for x and y that
```

File: examples/api/cpp/quickstart.cpp [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+  std::string xStr = xVal.getRealValue();
+  std::string yStr = yVal.getRealValue();
+
+  // now we can convert the values to cpp types
+  // using standard cpp conversion functions.
+  double xDouble = std::stod(xStr);
+  double yDouble = std::stod(yStr);
+
+  std::cout << "solution for x: " << xDouble << std::endl;
+  std::cout << "solution for y: " << yDouble << std::endl;
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:121]: You could just print `xStr` and `yStr` without having to convert them to double, no?

File: examples/api/cpp/quickstart.cpp [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+  // Next, we will check satisfiability of the same formula,
+  // only this time over integer variables a,b
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:124]:
```suggestion
  // only this time over integer variables a and b.
```

File: examples/api/cpp/quickstart.cpp [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+  // We start by reseting the solver to flush
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:126]:
```suggestion
  // We start by resetting assertions added to the solver.
```

File: examples/api/cpp/quickstart.cpp [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+  // previous assertions.
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:127]:
```suggestion
```

File: examples/api/cpp/quickstart.cpp [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+  solver.resetAssertions();
+
+  // Next, we assert the same assertions above with integers,
+  // in an abbreviated manner.
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:131]: What is the abbreviated manner?

File: examples/api/cpp/quickstart.cpp [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+  solver.assertFormula(solver.mkTerm(LT, solver.mkInteger(0), a));
+  solver.assertFormula(solver.mkTerm(LT, solver.mkInteger(0), b));
+  solver.assertFormula(
+      solver.mkTerm(LT, solver.mkTerm(PLUS, a, b), solver.mkInteger(1)));
+  solver.assertFormula(solver.mkTerm(LEQ, a, b));
+  // We check whether the revised assertion is satisfiable.
+  Result r2 = solver.checkSat();
+
+  // This time the formula is unsatisfiable
+  std::cout << "result: " << r2 << std::endl;
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:141]:
```suggestion
  std::cout << "expected: unsat" << std::endl;
  std::cout << "result: " << r2 << std::endl;
```

**Review by aniemetz — COMMENTED** [review]
Overall, the example looks really good.
What we might want to add:

* get value of an expression that is not part of an assertion, e.g., x * y
* create an indexed term, maybe? maybe this is out of scope for this one, though

Can you also maybe rewrap the comments a bit to better make use of the 80 char length? Some are broken after half of the line.

**Comment by yoni206** [comment]
Thanks! I added a `get-value` for `x-y` (which is not in the original formula).

Regarding indexed operator, personally I wouldn't put them in the very first example, but I don't feel strongly about it so let me know if you prefer I add these.

Regarding the conversion to `double` before printing, I now added some lines that manipulate the `double`s (my goal was to show how to get actuall arithmetic values and not strings). 

All other comments are addressed.

File: examples/api/cpp/quickstart.cpp [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+  // Another way to independently compute the value of x and y would be using
+  // ordinary cpp minus operator, rather than asking the solver.
+  // However, for more complex terms,
+  // it is easier to let the solver do the evaluation.
+  double xMinusYComputed = xDouble - yDouble;
+  if (xMinusYComputed == xMinusYDouble) {
+    std::cout << "computed correctly" << std::endl;
+  } else {
+    std::cout << "computed incorrectly" << std::endl;
+  }
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:132]: Format.

File: examples/api/cpp/quickstart.cpp [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+  std::cout << "value for x - y: " << xMinusYStr << std::endl;
+
+  // Further, we can convert the values to cpp types,
+  // using standard cpp conversion functions.
+  double xDouble = std::stod(xStr);
+  double yDouble = std::stod(yStr);
+  double xMinusYDouble = std::stod(xMinusYStr);
+
+  // Another way to independently compute the value of x and y would be using
+  // ordinary cpp minus operator, rather than asking the solver.
```
Comment (aniemetz) [inline examples/api/cpp/quickstart.cpp:124]:
```suggestion
  // the ordinary cpp minus operator instead of asking the solver.
```

**Review by aniemetz — COMMENTED** [review]
Just two minor things, else LGTM.

**Review by aniemetz — APPROVED** [review]

**Pull request merged**